src/utils/hooks/useChangedQuerySettingsIndicator.ts

```diff
@@ -0,0 +1,71 @@
+import React from 'react';
+
+import getChangedQueryExecutionSettings from '../../containers/Tenant/Query/QueryEditorControls/utils/getChangedQueryExecutionSettings';
+import getChangedQueryExecutionSettingsDescription from '../../containers/Tenant/Query/QueryEditorControls/utils/getChangedQueryExecutionSettingsDescription';
+import {
+    DEFAULT_QUERY_SETTINGS,
+    QUERY_SETTINGS_BANNER_LAST_CLOSED_KEY,
+    WEEK_IN_SECONDS,
+} from '../constants';
+
+import {useLastQueryExecutionSettings} from './useLastQueryExecutionSettings';
+import {useSetting} from './useSetting';
+
+export const useChangedQuerySettingsIndicator = () => {
+    const [bannerLastClosedTimestamp, setBannerLastClosedTimestamp] = useSetting<
+        number | undefined
+    >(QUERY_SETTINGS_BANNER_LAST_CLOSED_KEY);
+    const [isBannerShown, setIsBannerShown] = React.useState(false);
+    const [isIndicatorShown, setIsIndicatorShown] = React.useState(false);
+    const [changedSettings, setChangedSettings] = React.useState<string[]>([]);
+    const [lastQuerySettings] = useLastQueryExecutionSettings();
+    const changedSettingsArray = React.useMemo(() => {
+        return lastQuerySettings
+            ? getChangedQueryExecutionSettings(lastQuerySettings, DEFAULT_QUERY_SETTINGS)
+            : [];
+    }, [lastQuerySettings]);
+
+    const changedSettingsDescription = React.useMemo(() => {
+        return lastQuerySettings
+            ? getChangedQueryExecutionSettingsDescription({
+                  currentSettings: lastQuerySettings,
+                  defaultSettings: DEFAULT_QUERY_SETTINGS,
+              })
+            : '';
+    }, [lastQuerySettings]);
+
+    const checkIndicatorVisibility = React.useCallback(() => {
```

Comment:
is it possible to get rid of this `useEffect` and calculate all values during rendering cycle?

Reply:
sure!